java-squid/src/main/java/org/sonar/java/se/AbstractSymbolicValueRelation.java

```diff
@@ -0,0 +1,166 @@
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.java.se;
+
+import javax.annotation.CheckForNull;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public abstract class AbstractSymbolicValueRelation {
+
+  protected final SymbolicValue v1;
+  protected final SymbolicValue v2;
+
+  protected AbstractSymbolicValueRelation(SymbolicValue v1, SymbolicValue v2) {
+    this.v1 = v1;
+    this.v2 = v2;
+  }
+
+  @Override
+  public int hashCode() {
+    long bits = v1.id();
+    bits ^= v2.id() * 31;
+    bits ^= getClass().hashCode() * 31;
+    return ((int) bits) ^ ((int) (bits >> 32));
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    if (obj instanceof AbstractSymbolicValueRelation) {
+      return equalsRelation((AbstractSymbolicValueRelation) obj);
+    }
+    return super.equals(obj);
+  }
+
+  private boolean equalsRelation(AbstractSymbolicValueRelation relation) {
+    final int[] key1 = key();
+    final int[] key2 = relation.key();
+    if (key1[0] == key2[0] && key1[1] == key2[1]) {
+      return getClass().equals(relation.getClass());
+    }
+    return false;
+  }
+
+  private int[] key() {
+    if (v1.id() < v2.id()) {
+      return new int[] {v1.id(), v2.id()};
+    } else {
+      return new int[] {v2.id(), v1.id()};
+    }
+  }
+
+  @CheckForNull
+  public Boolean impliedBy(Collection<AbstractSymbolicValueRelation> knownRelations) {
+    return impliedBy(knownRelations, new HashSet<AbstractSymbolicValueRelation>());
+  }
+
+  @CheckForNull
+  public Boolean impliedBy(Collection<AbstractSymbolicValueRelation> knownRelations, Set<AbstractSymbolicValueRelation> usedRelations) {
+    if (knownRelations.isEmpty()) {
+      return null;
+    }
+    if (usedRelations.size() > 200) {
+      throw new IllegalStateException();
+    }
+    for (AbstractSymbolicValueRelation relation : knownRelations) {
+      Boolean result = relation.implies(this);
+      if (result != null) {
+        return result;
+      }
+      usedRelations.add(relation);
+      usedRelations.add(relation.symmetric());
+    }
+    return impliedBy(transitiveReduction(knownRelations, usedRelations), usedRelations);
+  }
+
+  private Collection<AbstractSymbolicValueRelation> transitiveReduction(Collection<AbstractSymbolicValueRelation> knownRelations, Set<AbstractSymbolicValueRelation> usedRelations) {
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Split this 181 characters long line (which is greater than 180 authorized). [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS00103)